File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool ok(TreeNode* root, long long low, long long high){
        if (!root) return true;
        if (root->val <= low || root->val >= high)
            return false;
        return ok(root->left,low,root->val) && ok(root->right,root->val,high);
    }
    bool isValidBST(TreeNode* root) {
        return ok(root,LLONG_MIN,LLONG_MAX);
    }
};

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        MIN = -2**31-1
        MAX = 2**31+2
        def isValid(root: Optional[TreeNode],MIN: int,MAX: int) -> bool:
            if not root:
                return True
            if root.val <= MIN or root.val >= MAX:
                return False
            return isValid(root.left,MIN,root.val) and isValid(root.right,root.val,MAX)
        return isValid(root,MIN,MAX)